### PR TITLE
Replication plug

### PR DESCRIPTION
**Pull request by aafemt**

### CODE REVIEW

File: src/jrd/replication/Replicator.h [hunk 1]
```diff
@@ -225,8 +225,16 @@ namespace Replication
 				   const Firebird::MetaString& userName,
 				   bool cleanupTransactions);
 
-		// IDisposable methods
-		void dispose();
+		// IeferenceCounted methods
```
Comment (dyemanov) [inline src/jrd/replication/Replicator.h:228]: Typo here -> IeferenceCounted

File: src/include/firebird/FirebirdInterface.idl [hunk 2]
```diff
@@ -1557,16 +1558,39 @@ interface Int128 : Versioned
...
+	// Return false if field value is not NULL, true otherwise.
+	boolean isNull(uint index);
+	// Return scale fo the field. Behavior is undefined for non-numeric fields.
+	uint getScale(uint index);
+	// Return maximum length of data in bytes. 0 is returned on error.
+	uint getLength(uint index);
+	// Return character set without collation part for text or CLOB fields. Behavior is undefined for the rest.
+	uint getCharSet(uint index);
+	// Return pointer to data. NULL is returned on error.
+	const void* getData(uint index);
```
Comment (dyemanov) [inline src/include/firebird/FirebirdInterface.idl:1578]: With the ReplicatedField interface, getData() may return NULL if the value is NULL. isNull() method becomes unnecessary.

**Review by dyemanov — CHANGES_REQUESTED** [review]

File: builds/install/misc/replication.conf [hunk 3]
```diff
@@ -6,6 +6,10 @@ database
 {
 	### ORIGIN SIDE SETTINGS
 
+	# Plugin used to perform replication.
+	#
+	# plugin = builtin
+
```
[the remark this answers is not in the document]
Comment (aafemt) [reply builds/install/misc/replication.conf:12]: I would prefer empty name for disabling replication. Now enabling it requires setting of log_directory that is pointless for external plugins (they have no access to Replication::Config).

File: src/include/firebird/FirebirdInterface.idl [hunk 4]
```diff
@@ -1586,17 +1610,17 @@ interface ReplicatedTransaction : Disposable
...
-interface ReplicatedSession : Disposable
+interface ReplicatedSession : PluginBase
 {
 	Status getStatus();
 
-	ReplicatedTransaction startTransaction(int64 number);
+	// Parameter don't need to be released if not used.
+	void setAttachment(Attachment attachment);
+	ReplicatedTransaction startTransaction(Transaction transaction, int64 number);
 	boolean cleanupTransaction(int64 number);
```
[the remark this answers is not in the document]
Comment (aafemt) [reply src/include/firebird/FirebirdInterface.idl:1624]: A lot of usage up to performing some queries to system or plugin defined-tables.

File: src/jrd/replication/Publisher.cpp [hunk 5]
```diff
@@ -102,18 +98,33 @@ namespace
...
+					logOriginMessage(dbb->dbb_filename, msg, ERROR_MSG);
+					return nullptr;
+				}
+				attachment->att_replicator = plugins.plugin();
+			}
 
-		fb_assert(attachment->att_replicator);
+			if (attachment->att_replicator.hasData())
+				attachment->att_replicator->setAttachment(attachment->getInterface());
+		}
```
[the remark this answers is not in the document]
Comment (aafemt) [reply src/jrd/replication/Publisher.cpp:123]: Plugin factory returning nullptr.

File: src/jrd/replication/Publisher.cpp [hunk 6]
```diff
@@ -102,18 +98,33 @@ namespace
...
 
-		// Disable replication after errors
+		fb_assert(attachment->att_replicator.hasData());
 
-		const auto status = attachment->att_replicator->getStatus();
-		if (status->getState() & IStatus::STATE_ERRORS)
-			return NULL;
+		// No need to check for errors here: fresh replicator cannot have any
```
[the remark this answers is not in the document]
Comment (aafemt) [reply src/jrd/replication/Publisher.cpp:128]: Yes and it can have some old (already handled) errors in status because you decided it to be static. It would be bad to react to them here.

File: builds/install/misc/replication.conf (hunk 3 above)
[the remark this answers is not in the document]
Comment (AlexPeshkoff) [reply builds/install/misc/replication.conf:12]: I agree with DS. In all other places when working with plugins we follow mentioned logic - empty plugins list disables feature, default is explicit. I see no problems with predefined name of plugin - we already have a lot in other places, starting with providers' names.

[the remark this answers is not in the document]
Comment (dyemanov) [reply builds/install/misc/replication.conf:12]: Disabling is primarily done by excluding / commenting the database section of the replication.conf file. log_directory can be emptied to disable async replication, if required. Adding yet another level of enabled/disabled seems error-prone to me. But I'm not going to insist if others don't see that as over-complicated.

[the remark this answers is not in the document]
Comment (aafemt) [reply builds/install/misc/replication.conf:12]: > Disabling is primarily done by excluding / commenting the database section of the replication.conf file.

Actually it should be done by "ALTER DATABASE DISABLE PUBLICATION" at first hand and here comes another problem: I see no way to enable replication of all (any) database in the configuration file, only every one individually. Have I missed something?

[the remark this answers is not in the document]
Comment (dyemanov) [reply builds/install/misc/replication.conf:12]: No, there's no such a way. Primarily because I cannot imagine multiple replicated databases sharing the same settings (first of all targets). While it's technically possible for them to share the log directories, it gonna be a maintenance nightmare for DBA. Perhaps we could match databases by patterns in replication.conf, but it requires serious thinking ( see above). Anyway, I'm open to ideas.

**Comment by AlexPeshkoff** [comment]
On 2020-08-24 19:27, Dmitry Yemanov wrote:
>
> *@dyemanov* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In builds/install/misc/replication.conf 
> <https://github.com/FirebirdSQL/firebird/pull/280#discussion_r475738286>:
>
> > @@ -6,6 +6,10 @@ database
>   {
>   	### ORIGIN SIDE SETTINGS
>   
> +	# Plugin used to perform replication.
> +	#
> +	# plugin = builtin
> +
>
> Disabling is primarily done by excluding / commenting the database 
> section of the replication.conf file.
>

Now I see - that's not list of plugins, available to perform some 
action, but particular plugin, configured for particular DB. In that 
case I see no problems with existing approach.

File: builds/install/misc/replication.conf (hunk 3 above)
[the remark this answers is not in the document]
Comment (aafemt) [reply builds/install/misc/replication.conf:12]: Even if setting is set for a pattern it still can use individual settings for each matched database as in the trace config:
```
database = /db/(*.fdb)
{
  sync_replica = server2:/replicas/\1
}
```

**Comment by aafemt** [comment]
> that's not list of plugins, available to perform some action, but particular plugin, configured for particular DB.

And as such it has no difference, for example, from db crypt plugin.
In a proper implementation all setting (including used plugin name) should go to publication table. But it would push release date even more so for now let's limit ourselves to something that is barely enough for two particular plugins.

File: builds/install/misc/replication.conf (hunk 3 above)
[the remark this answers is not in the document]
Comment (dyemanov) [reply builds/install/misc/replication.conf:12]: I'm afraid filename substitutions may be not enough. But I agree to consider that for v4.1.

File: src/include/firebird/FirebirdInterface.idl [hunk 7]
```diff
@@ -1557,16 +1558,39 @@ interface Int128 : Versioned
 
 interface ReplicatedRecord : Versioned
 {
+	// Return count of fields
+	uint getCount();
+	// Return name of field. NULL is returned on error.
+	const string getField(uint index);
+	// Return dtype_* constants or dtype_unknown if field not exists.
+	uint getType(uint index);
```
[the remark this answers is not in the document]
Comment (asfernandes) [reply src/include/firebird/FirebirdInterface.idl:1566]: Why storage internals should be exposed to replication plugin if message types can be correspondent to them?

AFAIK only place dtypes are exposed are for UDFs and they are deprecated. I do not see a sense in reinsert them as public thing.

[the remark this answers is not in the document]
Comment (aafemt) [reply src/include/firebird/FirebirdInterface.idl:1566]: Nope, they are exposed to public in dsc_pub.h which is included into ibase.h and Interface.h.

[the remark this answers is not in the document]
Comment (asfernandes) [reply src/include/firebird/FirebirdInterface.idl:1566]: Exposed for what usage currently?

[the remark this answers is not in the document]
Comment (aafemt) [reply src/include/firebird/FirebirdInterface.idl:1566]: Frankly, I have no idea. It is too old thing.

[the remark this answers is not in the document]
Comment (asfernandes) [reply src/include/firebird/FirebirdInterface.idl:1566]: So you confirm my point it should not continue to be exposed as there is something that already replaces it externally.

[the remark this answers is not in the document]
Comment (dyemanov) [reply src/include/firebird/FirebirdInterface.idl:1566]: I considered that, but found it inappropriate. Record format has to be repacked into a message format for every record modification and this may be a noticeable performance issue. It would be great to have messages more flexible (so that they could be accessed from different storage formats using the same API), but currently it's not really so. Perhaps field values could be accessed using getOffset() returning an internal record position (do we have the same alignment rules for record fields and message elements?), but getNullOffset() breaks everything -- representing NULLs as SHORT is absolutely dumb requirement that's inherited from the legacy code. Calls like getRelation() / getOwner() / getAlias() are also redundant here, but that's not a big deal - empty string can be returned.

[the remark this answers is not in the document]
Comment (dyemanov) [reply src/include/firebird/FirebirdInterface.idl:1566]: That said, I have no objections for getType() returning the SQL type rather than dtype (for consistency with other API).

[the remark this answers is not in the document]
Comment (aafemt) [reply src/include/firebird/FirebirdInterface.idl:1566]: >  I have no objections for getType() returning the SQL type rather than dtype

So do I.

Another question: should getLength() also follow this rule? The length of dtype_varying is two bytes bigger that length of SQL_VARYING for the same field.

[the remark this answers is not in the document]
Comment (dyemanov) [reply src/include/firebird/FirebirdInterface.idl:1566]: I think yes, it should. If we return SQL_VARYING as a type, then its size should also correspond to SQL_VARYING.

File: builds/install/misc/replication.conf (hunk 3 above)
[the remark this answers is not in the document]
Comment (aafemt) [reply builds/install/misc/replication.conf:12]: Is it too late to add field rdb$plugin_name into table rdb$publications?

File: src/include/firebird/FirebirdInterface.idl (hunk 7 above)
[the remark this answers is not in the document]
Comment (asfernandes) [reply src/include/firebird/FirebirdInterface.idl:1566]: > That said, I have no objections for getType() returning the SQL type rather than dtype (for consistency with other API).

That is what I was suggesting mainly.

File: builds/install/misc/replication.conf (hunk 3 above)
[the remark this answers is not in the document]
Comment (dyemanov) [reply builds/install/misc/replication.conf:12]: Instead of "plugin" in replication.conf?

[the remark this answers is not in the document]
Comment (aafemt) [reply builds/install/misc/replication.conf:12]: Yes. There is no point to have the same setting in two places. In this case active publication in database will be a sufficient for initializing replication from engine side and it can be automatically disabled later if/when internal plugin detect other required settings missing.

[the remark this answers is not in the document]
Comment (dyemanov) [reply builds/install/misc/replication.conf:12]: Let's complete this PR with the config-based plugin setting. If adding a field would be possible after that (depends on whether some my changes will require ODS changes or not), it should be done with another PR.

**Comment by aafemt** [comment]
This time with requested changes and improved resource management.

**Comment by aafemt** [comment]
It looks like clang is better in resolving of names conflict than MSVC.

File: src/include/firebird/IdlFbInterfaces.h [hunk 8]
```diff
@@ -1990,7 +1991,7 @@ namespace Firebird
 
 		static const int EXECUTE_FAILED = -1;
 		static const int SUCCESS_NO_INFO = -2;
-		static const unsigned NO_MORE_ERRORS = 0xffffffff;
+		static const unsigned NO_MORE_ERRORS = 0x7fffffff;
```
Comment (dyemanov) [inline src/include/firebird/IdlFbInterfaces.h:1994]: I don't understand this change, it doesn't seem related to your changes inside the .idl file.

File: src/jrd/replication/Publisher.cpp [hunk 9]
```diff
@@ -226,74 +244,100 @@ namespace
...
+		}
+
+		IReplicatedField* getField(unsigned index) override
+		{
+			if (index >= m_format->fmt_count)
+				return nullptr;
+
+			m_fieldIndex = index;
+			SLONG dummySubtype, dummyScale;
+			m_format->fmt_desc[m_fieldIndex].getSqlInfo(&m_fieldLength, &dummySubtype, &dummyScale, &m_fieldType);
```
Comment (dyemanov) [inline src/jrd/replication/Publisher.cpp:318]: Why not including subtype and scale to the record object too? Stack allocation is dirty cheap and returning the value should be slightly faster than m_format->fmt_desc[m_fieldIndex] redirection + method call.

File: src/jrd/replication/Publisher.cpp [hunk 10]
```diff
@@ -303,21 +347,26 @@ void REPL_attach(thread_db* tdbb, bool cleanupTransactions)
...
-	if (!replMgr)
+	const auto replConfig = dbb->replConfig();
+	if (!replConfig)
 		return;
 
 	fb_assert(!attachment->att_repl_matcher);
 	auto& pool = *attachment->att_pool;
-	attachment->att_repl_matcher = replMgr->createTableMatcher(pool);
+	attachment->att_repl_matcher = FB_NEW_POOL(pool)
+				TableMatcher(pool, replConfig->includeFilter, replConfig->excludeFilter);
```
Comment (dyemanov) [inline src/jrd/replication/Publisher.cpp:358]: Do you intend to use include/exclude filters for custom plugins too? If/When plugin name is moved from the config file to RDB$PUBLICATIONS, then I'd expect that replication plugins don't need replication.conf at all (plugin may provide its own filters, if necessary).

Comment (dyemanov) [reply src/jrd/replication/Publisher.cpp:358]: If there's no such an intention, then the matcher does not make any sense before the manager is created. If you want to defer manager creation, then the matcher should be created inside getReplicator() together with the replicator itself. In this case, I see no sense to create the matcher from config rather than from the manager. And all the changes around replManager() and replConfig() look unnecessary.

File: src/jrd/replication/Publisher.cpp [hunk 11]
```diff
@@ -560,11 +560,12 @@ void REPL_erase(thread_db* tdbb, const record_param* rpb, jrd_tra* transaction)
 
 	// This temporary auto-pointer is just to delete a temporary record
 	AutoPtr<Record> cleanupRecord(record != rpb->rpb_record ? record : NULL);
+	AutoSetRestoreFlag<ULONG> noRecursion(&tdbb->tdbb_flags, TDBB_repl_sql, true);
```
Comment (dyemanov) [inline src/jrd/replication/Publisher.cpp:564]: What is the problem you're solving here? Plugin modifying tables/generators inside the database? Should we really consider this appropriate?

**Review by dyemanov — CHANGES_REQUESTED** [review]
Please post changes in separate commits instead of force-modifying the original one. It's annoying to review the whole patch again and again. We may squash the commits into a single one when merging this PR.

**Comment by aafemt** [comment]
> Please post changes in separate commits instead of force-modifying the original one.

Ok, I'll try to avoid it. This force push was unavoidable because of rebasing to the top of #272 which in turn was necessary to regenerate headers after IDL change.

File: src/include/firebird/IdlFbInterfaces.h (hunk 8 above)
Comment (aafemt) [reply src/include/firebird/IdlFbInterfaces.h:1994]: So do I. I have no idea why it is here.

File: src/jrd/replication/Publisher.cpp [hunk 12]
```diff
@@ -226,74 +244,100 @@ namespace
...
 		~ReplicatedRecordImpl()
 		{
 		}
 
-		unsigned getRawLength()
+		unsigned getCount() override
 		{
-			return m_record->getLength();
+			return m_format->fmt_count;
 		}
```
[the remark this answers is not in the document]
Comment (aafemt) [reply src/jrd/replication/Publisher.cpp:266]: It would require code in getField() scanning whole field list from the beginning on each call. It shall turns O(N) into O(N^2) for record data handling. Unacceptable price for convenience, IMHO.

File: src/jrd/replication/Publisher.cpp (hunk 9 above)
Comment (aafemt) [reply src/jrd/replication/Publisher.cpp:318]: They are involved in charset trick. Calling existing routines is better than copy-pasting them in this case.

File: src/jrd/replication/Publisher.cpp (hunk 10 above)
Comment (aafemt) [reply src/jrd/replication/Publisher.cpp:358]: I have no idea why you didn't remove the matcher after making including tables into publication in SQL so preserved your logic as much as possible.

If you wish, I'll drop the matchers completely.

Comment (aafemt) [reply src/jrd/replication/Publisher.cpp:358]: Most of changes around replManager() caused by fact that it is useful only for internal replication. Additional thread that is created inside is a waste of resources for any external plugin.

File: src/jrd/replication/Publisher.cpp (hunk 11 above)
Comment (aafemt) [reply src/jrd/replication/Publisher.cpp:564]: We cannot strictly prohibit it and the prevention is cheap so why not?

File: src/include/firebird/FirebirdInterface.idl [hunk 13]
```diff
@@ -1555,17 +1556,36 @@ interface Int128 : Versioned
...
 }
 
-interface ReplicatedBlob : Versioned
+interface ReplicatedRecord : Versioned
 {
-	uint getLength();
-	boolean isEof();
-	uint getSegment(uint length, uchar* buffer);
+	// Return count of fields
+	uint getCount();
```
[the remark this answers is not in the document]
Comment (aafemt) [reply src/include/firebird/FirebirdInterface.idl:1580]: In IMessageMetadata it is called getCount() I would prefer to be consistent.

[the remark this answers is not in the document]
Comment (dyemanov) [reply src/include/firebird/FirebirdInterface.idl:1580]: OK

File: src/jrd/replication/Publisher.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (dyemanov) [reply src/jrd/replication/Publisher.cpp:266]: I see no sense in returning something that cannot be used. As for performance, scanning could be done just once, in the constructor. A mapping table will be required, of course, but it could be made HalfStaticArray with large enough (e.g. 256) initial capacity and maybe even reused between multiple records via the Replicator object. Alternatively, scanning could be done per every field but starting with the prior m_fieldIndex (m_position should also be saved). The most common (sequential) field access pattern will perform with O(N).

File: src/jrd/replication/Publisher.cpp (hunk 11 above)
Comment (dyemanov) [reply src/jrd/replication/Publisher.cpp:564]: OK, but please rename the flag then. TDBB_replicating or TDBB_repl_in_progress or something like that.

File: src/jrd/replication/Publisher.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (aafemt) [reply src/jrd/replication/Publisher.cpp:266]: I still think that this is a needless complication. Check for getField() returning nullptr has to be in every plugin in any case so skipping of disappeared fields is easy there. At least in my plugin I would prefer to have a real field id (just in case if I decide to use it for fields mapping instead of name).

File: src/include/firebird/IdlFbInterfaces.h (hunk 8 above)
Comment (aafemt) [reply src/include/firebird/IdlFbInterfaces.h:1994]: Perhaps @asfernandes should look at it. CLOOP is his pet.

File: src/jrd/replication/Publisher.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (aafemt) [reply src/jrd/replication/Publisher.cpp:266]: Actually, now, when I said it, I realized that my existing code using field names will fail on renamed fields and that's a problem...

File: src/include/firebird/IdlFbInterfaces.h (hunk 8 above)
Comment (asfernandes) [reply src/include/firebird/IdlFbInterfaces.h:1994]: > Perhaps @asfernandes should look at it. CLOOP is his pet.

If you have any reproducible test case sent it to Firebird developers.

If you do not like any piece of code there, you can remove and reimplement them in your side products copied from Firebird.

File: src/jrd/replication/Publisher.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (dyemanov) [reply src/jrd/replication/Publisher.cpp:266]: OK, I can live with that. It would be up to plugin to handle missing fields.

File: src/jrd/replication/Publisher.cpp (hunk 10 above)
Comment (dyemanov) [reply src/jrd/replication/Publisher.cpp:358]: Regexps are handy if you need to exclude a subset of tables, so I decided to leave the matcher to have an override option (second level of customization). So far I don't want to drop it.

**Review by dyemanov — COMMENTED** [review]
I don't have problems with the current state of the PR. Once the TDBB flag is renamed, I'm ready to merge it. If something wrong / inappropriate will be found after the merge, I will fix it inside the repo myself.

**Comment by aafemt** [comment]
This PR should include replication error handling before being ready for merge.

File: src/jrd/replication/Publisher.cpp (hunk 10 above)
Comment (aafemt) [reply src/jrd/replication/Publisher.cpp:358]: I agree that regexps are handy. They should be supported in ALTER PUBLICATION statement.

File: src/include/firebird/IdlFbInterfaces.h (hunk 8 above)
Comment (aafemt) [reply src/include/firebird/IdlFbInterfaces.h:1994]: How is it related to obviously wrongly generated line above?

File: src/include/firebird/IdlFbInterfaces.h [hunk 14]
```diff
@@ -6235,17 +6272,22 @@ namespace Firebird
 		}
 
 	public:
-		static const unsigned VERSION = 3;
+		static const unsigned VERSION = 4;
```
Comment (aafemt) [inline src/include/firebird/IdlFbInterfaces.h:6275]: I wonder why version of ReplicatedSession has been changed while version of ReplicatedRecord wasn't.

**Comment by aafemt** [comment]
Now it is complete.

**Comment by asfernandes** [comment]
> Now it is complete.

Looks like not. Build still fail.

**Comment by aafemt** [comment]
> Build still fail.

Just because of unmerged #282

File: src/jrd/replication/Publisher.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (dyemanov) [reply src/jrd/replication/Publisher.cpp:152]: Given the condition if (!plugins.hasData()) above, the factory cannot return nullptr, AFAIU.

[the remark this answers is not in the document]
Comment (aafemt) [reply src/jrd/replication/Publisher.cpp:152]: Yes, you are right.
But still internal replication plugin creation can be skipped because of missing configuration settings ending up in att_replicator containing nullptr so the check here is needed bu assert below is better to be removed.
